### ISSUE TITLE
Crash on first run with local backend.

### ISSUE THREAD

**Issue by akvadrako** [greenpau/caddy-auth-portal#97]
**Describe the issue**

When I start caddy for the first time without a `users.json` file, it crashes. Starting it a second time and each subsequent time works.

```
Attaching to auth_auth_1
auth_1  | {"level":"info","ts":1614956846.1614413,"msg":"using provided configuration","config_file":"conf/Caddyfile","config_adapter":""}
auth_1  | {"level":"info","ts":1614956846.1634688,"logger":"admin","msg":"admin endpoint started","address":"tcp/localhost:2019","enforce_origin":false,"origins":["localhost:2019","[::1]:2019","127.0.0.1:2019"]}
auth_1  | {"level":"info","ts":1614956846.1638024,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0xc0001db110"}
auth_1  | {"level":"info","ts":1614956846.1639771,"logger":"http.authentication.providers.jwt","msg":"provisioned plugin instance","instance_name":"jwt-1","started_at":1614956846.1638982}
auth_1  | {"level":"info","ts":1614956846.1642427,"logger":"http.handlers.auth_portal","msg":"JWT token name found","instance_name":"portal-1","token_name":"access_token"}
auth_1  | {"level":"warn","ts":1614956846.164379,"logger":"http.handlers.auth_portal","msg":"JWT token origin not found, using default","instance_name":"portal-1"}
auth_1  | {"level":"info","ts":1614956846.1643906,"logger":"http.handlers.auth_portal","msg":"local backend configuration","db_path":"/home/data/users.json","require_mfa":true}
auth_1  | {"level":"info","ts":1614956846.1644037,"logger":"http.handlers.auth_portal","msg":"local database file does not exists, creating it","db_path":"/home/data/users.json"}
auth_1  | panic: runtime error: invalid memory address or nil pointer dereference
auth_1  | [signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x149b9ad]
auth_1  |
auth_1  | goroutine 1 [running]:
auth_1  | github.com/greenpau/caddy-auth-portal/pkg/backends/local.(*Authenticator).Configure(0xc0008651a0, 0x0, 0x0)
auth_1  | 	github.com/greenpau/caddy-auth-portal@v1.4.4-0.20210220075436-1d42614c6a75/pkg/backends/local/backend.go:142 +0x28d
auth_1  | github.com/greenpau/caddy-auth-portal/pkg/backends/local.(*Backend).ConfigureAuthenticator(0xc000861c80, 0xc0010b20e0, 0xc0010b1c90)
auth_1  | 	github.com/greenpau/caddy-auth-portal@v1.4.4-0.20210220075436-1d42614c6a75/pkg/backends/local/backend.go:275 +0x9d
auth_1  | github.com/greenpau/caddy-auth-portal/pkg/backends.(*Backend).Configure(0xc0010b2050, 0xc0010b20e0, 0x18e78bb, 0xe)
auth_1  | 	github.com/greenpau/caddy-auth-portal@v1.4.4-0.20210220075436-1d42614c6a75/pkg/backends/backend.go:99 +0x3ff
auth_1  | github.com/greenpau/caddy-auth-portal/pkg/core.(*AuthPortalManager).Register(0xc0006ac700, 0xc0001b3380, 0x0, 0x0)
auth_1  | 	github.com/greenpau/caddy-auth-portal@v1.4.4-0.20210220075436-1d42614c6a75/pkg/core/manager.go:231 +0x5705
auth_1  | github.com/greenpau/caddy-auth-portal/pkg/core.(*AuthPortal).Configure(0xc0001b3380, 0xc0010b3848, 0x18b5c3c, 0x6)
auth_1  | 	github.com/greenpau/caddy-auth-portal@v1.4.4-0.20210220075436-1d42614c6a75/pkg/core/portal.go:98 +0x1bd
auth_1  | github.com/greenpau/caddy-auth-portal.(*AuthMiddleware).Provision(0xc000121220, 0x25f4f30, 0xc00075ea80, 0xc000813590, 0xc00080c370, 0x0, 0x0, 0x0, 0x455454, 0x19)
auth_1  | 	github.com/greenpau/caddy-auth-portal@v1.4.4-0.20210220075436-1d42614c6a75/plugin.go:48 +0x166
auth_1  | github.com/caddyserver/caddy/v2.Context.LoadModuleByID(0x25f4f30, 0xc00075ea80, 0xc000813590, 0xc00080c370, 0x0, 0x0, 0x0, 0xc000465900, 0x19, 0xc000187500, ...)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/context.go:344 +0x689
auth_1  | github.com/caddyserver/caddy/v2.Context.loadModuleInline(0x25f4f30, 0xc00075ea80, 0xc000813590, 0xc00080c370, 0x0, 0x0, 0x0, 0x16ba480, 0x7, 0x16ba467, ...)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/context.go:394 +0x187
auth_1  | github.com/caddyserver/caddy/v2.Context.LoadModule(0x25f4f30, 0xc00075ea80, 0xc000813590, 0xc00080c370, 0x0, 0x0, 0x0, 0x16967e0, 0xc00023db00, 0x18cf057, ...)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/context.go:175 +0x86f
auth_1  | github.com/caddyserver/caddy/v2/modules/caddyhttp.RouteList.ProvisionHandlers(0xc00023db00, 0x1, 0x4, 0x25f4f30, 0xc00075ea80, 0xc000813590, 0xc00080c370, 0x0, 0x0, 0x0, ...)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/modules/caddyhttp/routes.go:150 +0xce
auth_1  | github.com/caddyserver/caddy/v2/modules/caddyhttp.RouteList.Provision(0xc00023db00, 0x1, 0x4, 0x25f4f30, 0xc00075ea80, 0xc000813590, 0xc00080c370, 0x0, 0x0, 0x0, ...)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/modules/caddyhttp/routes.go:122 +0xff
auth_1  | github.com/caddyserver/caddy/v2/modules/caddyhttp.(*Subroute).Provision(0xc000859900, 0x25f4f30, 0xc00075ea80, 0xc000813590, 0xc00080c370, 0x0, 0x0, 0x0, 0x455454, 0x16)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/modules/caddyhttp/subroute.go:58 +0x95
auth_1  | github.com/caddyserver/caddy/v2.Context.LoadModuleByID(0x25f4f30, 0xc00075ea80, 0xc000813590, 0xc00080c370, 0x0, 0x0, 0x0, 0xc000125950, 0x16, 0xc000808380, ...)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/context.go:344 +0x689
auth_1  | github.com/caddyserver/caddy/v2.Context.loadModuleInline(0x25f4f30, 0xc00075ea80, 0xc000813590, 0xc00080c370, 0x0, 0x0, 0x0, 0x16ba480, 0x7, 0x16ba467, ...)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/context.go:394 +0x187
auth_1  | github.com/caddyserver/caddy/v2.Context.LoadModule(0x25f4f30, 0xc00075ea80, 0xc000813590, 0xc00080c370, 0x0, 0x0, 0x0, 0x16967e0, 0xc00023d8c0, 0x18cf057, ...)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/context.go:175 +0x86f
auth_1  | github.com/caddyserver/caddy/v2/modules/caddyhttp.RouteList.ProvisionHandlers(0xc00023d8c0, 0x1, 0x4, 0x25f4f30, 0xc00075ea80, 0xc000813590, 0xc00080c370, 0x0, 0x0, 0x0, ...)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/modules/caddyhttp/routes.go:150 +0xce
auth_1  | github.com/caddyserver/caddy/v2/modules/caddyhttp.RouteList.Provision(0xc00023d8c0, 0x1, 0x4, 0x25f4f30, 0xc00075ea80, 0xc000813590, 0xc00080c370, 0x0, 0x0, 0x0, ...)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/modules/caddyhttp/routes.go:122 +0xff
auth_1  | github.com/caddyserver/caddy/v2/modules/caddyhttp.(*Subroute).Provision(0xc0008597c0, 0x25f4f30, 0xc00075ea80, 0xc000813590, 0xc00080c370, 0x0, 0x0, 0x0, 0x455454, 0x16)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/modules/caddyhttp/subroute.go:58 +0x95
auth_1  | github.com/caddyserver/caddy/v2.Context.LoadModuleByID(0x25f4f30, 0xc00075ea80, 0xc000813590, 0xc00080c370, 0x0, 0x0, 0x0, 0xc0001258d8, 0x16, 0xc00061ee00, ...)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/context.go:344 +0x689
auth_1  | github.com/caddyserver/caddy/v2.Context.loadModuleInline(0x25f4f30, 0xc00075ea80, 0xc000813590, 0xc00080c370, 0x0, 0x0, 0x0, 0x16ba480, 0x7, 0x16ba467, ...)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/context.go:394 +0x187
auth_1  | github.com/caddyserver/caddy/v2.Context.LoadModule(0x25f4f30, 0xc00075ea80, 0xc000813590, 0xc00080c370, 0x0, 0x0, 0x0, 0x16967e0, 0xc00023d4d0, 0x18cf057, ...)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/context.go:175 +0x86f
auth_1  | github.com/caddyserver/caddy/v2/modules/caddyhttp.RouteList.ProvisionHandlers(0xc00023d440, 0x2, 0x4, 0x25f4f30, 0xc00075ea80, 0xc000813590, 0xc00080c370, 0x0, 0x0, 0x0, ...)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/modules/caddyhttp/routes.go:150 +0xce
auth_1  | github.com/caddyserver/caddy/v2/modules/caddyhttp.(*App).Provision(0xc000115c70, 0x25f4f30, 0xc00075ea80, 0xc000813590, 0xc00080c370, 0x0, 0x0, 0x0, 0xc000614700, 0x1)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/modules/caddyhttp/app.go:231 +0xb3c
auth_1  | github.com/caddyserver/caddy/v2.Context.LoadModuleByID(0x25f4f30, 0xc00075ea80, 0xc000813590, 0xc00080c370, 0x0, 0x0, 0x0, 0xc00065d4c0, 0x4, 0xc000302600, ...)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/context.go:344 +0x689
auth_1  | github.com/caddyserver/caddy/v2.Context.App(0x25f4f30, 0xc00075ea80, 0xc000813590, 0xc00080c370, 0x0, 0x0, 0x0, 0xc00065d4c0, 0x4, 0xc00075ea01, ...)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/context.go:415 +0x168
auth_1  | github.com/caddyserver/caddy/v2.run.func3(0xc00080c370, 0x25f4f30, 0xc00075ea80, 0xc000813590, 0xc00080c370, 0x0, 0x0, 0x0, 0x0, 0x0)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/caddy.go:388 +0xdc
auth_1  | github.com/caddyserver/caddy/v2.run(0xc00080c370, 0x601, 0x0, 0x0)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/caddy.go:393 +0x2ee
auth_1  | github.com/caddyserver/caddy/v2.unsyncedDecodeAndRun(0xc000833800, 0x6ab, 0x700, 0x7, 0xc0006a9e60)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/caddy.go:249 +0xff
auth_1  | github.com/caddyserver/caddy/v2.changeConfig(0x18adeb2, 0x4, 0x18b8333, 0x7, 0xc000832300, 0x6ab, 0x700, 0x7ffc909a6f01, 0x0, 0x0)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/caddy.go:156 +0x4df
auth_1  | github.com/caddyserver/caddy/v2.Load(...)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/caddy.go:102
auth_1  | github.com/caddyserver/caddy/v2/cmd.cmdRun(0xc000200600, 0x0, 0x0, 0x0)
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/cmd/commandfuncs.go:212 +0x3af
auth_1  | github.com/caddyserver/caddy/v2/cmd.Main()
auth_1  | 	github.com/caddyserver/caddy/v2@v2.3.0/cmd/main.go:85 +0x25b
auth_1  | main.main()
auth_1  | 	caddy/main.go:14 +0x25
auth_auth_1 exited with code 2
```

**Configuration**

```
{
  auto_https off
}

:80

log {
  level DEBUG
  format console
}

route /authtest {
  trace
  jwt {
    primary yes
    trusted_tokens {
      public_key {
        token_name access_token
        token_rsa_file 1 /home/data/public-rsa.pem
      }
    }
    auth_url /
    allow roles anonymous guest admin
  }
  respond * "authtest" 200
}

route {
  auth_portal {
    path /
    backends {
        local_backend {
            method local
            path /home/data/users.json
            realm local
            require mfa
        }
    }
    jwt {
        token_lifetime 3600
        token_rsa_file 1 /home/data/private-rsa.key
    }
    ui {
      links {
        "authtest" /authtest
        "Who am I?" /whoami target_blank icon "las la-star"
        "My Settings" /settings icon "las la-cog"
      }
    }
  }
}
```

**Version Information**

```
http.authentication.hashes.bcrypt v2.3.0
http.authentication.hashes.scrypt v2.3.0
http.authentication.providers.http_basic v2.3.0
http.authentication.providers.jwt v1.2.6
http.handlers.auth_portal v1.4.4-0.20210220075436-1d42614c6a75
http.handlers.authentication v2.3.0
```

**Expected behavior**

Do not crash.

**Comment by greenpau** [comment]
@akvadrako , just to confirm. The crash occurs only when there is no users.do file?

**Comment by greenpau** [comment]
Also I don’t think the plug-in works without https.
I was planning adding “insecure” directive to allow less secure cookies.

**Comment by akvadrako** [comment]
> @akvadrako , just to confirm. The crash occurs only when there is no users.do file?

Right, no `users.json`.

> Also I don’t think the plug-in works without https.

I am using it with https, it's just behind another reverse proxy. I think when testing locally it worked though, which is the only time I wouldn't use https.

**Comment by greenpau** [comment]
> I am using it with https, it's just behind another reverse proxy. I think when testing locally it worked though, which is the only time I wouldn't use https.

@akvadrako , yes :-) locally it might work. This is the line I was referring to:

https://github.com/greenpau/caddy-auth-portal/blob/1d42614c6a75b6a03fbcc7738500a3f06292ed55/pkg/cookies/cookie.go#L45

**Comment by greenpau** [comment]
@akvadrako , this is now resolved. Thank you for bringing it up.

When there is no database file, the plugin creates it and output username and password via info log.

```json
{
  "level": "info",
  "ts": 1615083448.0091398,
  "logger": "http.handlers.auth_portal",
  "msg": "created default superadmin user for the database",
  "user_name": "webadmin",
  "user_secret": "109b8431-d38e-4ec9-8f6d-f46d5769848e"
}
```

After, the admin could change the password.

The fix is now part of [v1.4.5](https://github.com/greenpau/caddy-auth-portal/releases/tag/v1.4.5)
